sw/cheri/common/sonata_plic.hh

```diff
@@ -0,0 +1,185 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "../../common/defs.h"
+#include <cheri.hh>
+
+namespace PLIC{
+ typedef enum : uint32_t {
+    // Uart 0 Interrupts.
+    Uart0TransmitWaterMark = 1,
+    Uart0ReceiveWaterMark = 2,
+    Uart0TransmitEmpty = 3,
+    Uart0ReceiveOverflow = 4,
+    Uart0ReceiveFrameError = 5,
+    Uart0ReceiveBreakError = 6,
+    Uart0ReceiveTimeout = 7,
+    Uart0ReceiveParityError = 8,
+    // Uart 1 Interrupts.
+    Uart1TransmitWaterMark = 9,
+    Uart1ReceiveWaterMark = 10,
+    Uart1TransmitEmpty = 11,
+    Uart1ReceiveOverflow = 12,
+    Uart1ReceiveFrameError = 13,
+    Uart1ReceiveBreakError = 14,
+    Uart1ReceiveTimeout = 15,
+    Uart1ReceiveParityError = 16,
+    // I2c 0 Interrupts.
+    I2c0FormatFifoThreshold = 17,
+    I2c0ReceiveFifoThreshold = 18,
+    I2c0AcquireFifoThreshold = 19,
+    I2c0ReceiveFifoOverflow = 20,
+    I2c0ReceiveNack = 21,
+    I2c0SclInterference = 22,
+    I2c0SdaInterference = 23,
+    I2c0StretchTimeout = 24,
+    I2c0SdaUnstable = 25,
+    I2c0CommandComplete = 26,
+    I2c0TransmitStretch = 27,
+    I2c0TransmitThreshold = 28,
+    I2c0AcquireFifoFull = 29,
+    I2c0UnexpectedStop = 30,
+    I2c0HostTimeout = 31,
+    // I2c 1 Interrupts.
+    I2c1FormatFifoThreshold = 32,
+    I2c1ReceiveFifoThreshold = 33,
+    I2c1AcquireFifoThreshold = 34,
+    I2c1ReceiveFifoOverflow = 35,
+    I2c1ReceiveNack = 36,
+    I2c1SclInterference = 37,
+    I2c1SdaInterference = 38,
+    I2c1StretchTimeout = 39,
+    I2c1SdaUnstable = 40,
+    I2c1CommandComplete = 41,
+    I2c1TransmitStretch = 42,
+    I2c1TransmitThreshold = 43,
+    I2c1AcquiteFifoFull = 44,
+    I2c1UnexpectedStop = 45,
+    I2c1HostTimeout = 46,
+    // Ethernet Interrupts.
+    Ethernet = 47,
+    MaxIntrID,
+ }Interrupts;
+
+ enum class Priority{
+    PriorityHigh,
+    PriorityMedium,
+    PriorityLow,
+  };
+
+  /**
+   * Driver for the standard RISC-V Platform-Local Interrupt Controller (PLIC).
+   */
+  class SonataPlic {
+    private:
+      // generic offsets according to spec
+      static constexpr size_t PriorityOffset  = 0x0U;
+      static constexpr size_t PendingOffset   = 0x1000U;
+      static constexpr size_t EnableOffset    = 0x2000U;
+      static constexpr size_t ThresholdOffset = 0x200000U;
+      static constexpr size_t ClaimOffset     = 0x200004U;
+
+      // Bounded capabilities to the individual structure fields.
+      // Ideally these should be contained in one volatile struct, but
+      // they are so far apart (and this lets us apply the bounds once).
+      volatile uint32_t *plicPrios;
+      volatile uint32_t *plicPendings;
+      volatile uint32_t *plicEnables;
+      volatile uint32_t *plicThres;
+      volatile uint32_t *plicClaim;
+
+    public:
+      /**
+       * Constructor. Initialises the interrupt controller with all interrupts disabled.
+       */
+      SonataPlic(CHERI::Capability<void> root ){
```

Comment:
I am not sure what causes this extra space?

Reply:
I rebased this PR and applied clang-format, should be fixed now